renderer/backend/UniversalGpuBuffer.cpp

```diff
@@ -0,0 +1,158 @@
+/*****************************************************************************
+                    The Dark Mod GPL Source Code
+
+ This file is part of the The Dark Mod Source Code, originally based
+ on the Doom 3 GPL Source Code as published in 2011.
+
+ The Dark Mod Source Code is free software: you can redistribute it
+ and/or modify it under the terms of the GNU General Public License as
+ published by the Free Software Foundation, either version 3 of the License,
+ or (at your option) any later version. For details, see LICENSE.TXT.
+
+ Project: The Dark Mod (http://www.thedarkmod.com/)
+
+******************************************************************************/
+#include "precompiled.h"
+#include "UniversalGpuBuffer.h"
+#include "../tr_local.h"
+
+extern idCVarBool r_usePersistentMapping;
+
+void UniversalGpuBuffer::Init( GLenum type, GLuint size, GLuint alignment ) {
+	if( mMapBase ) {
+		Destroy();
+	}
+
+	mSize = ALIGN( size, alignment );
+	mAlign = alignment;
+	mType = type;
+
+	qglGenBuffers( 1, &mBufferObject );
+	qglBindBuffer( type, mBufferObject );
+
+	mUsePersistentMapping = r_usePersistentMapping && glConfig.bufferStorageAvailable;
+
+	if (mUsePersistentMapping) {
+		qglBufferStorage( type, mSize, nullptr, GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
```

Comment:
At the moment, this class is used to store per drawcall params and multidraw commands for drawcall batching, so no static data.

Static data only becomes an issue if we want to migrate the VertexCache to also use this `UniversalGpuBuffer`. I think it would be a good idea, but this PR is already huge, so let's postpone that :)

And yeah, in general you'd want to use fewer buffers, but it's primarily about batching draw calls without state change - if static objects were still in a different buffer (before duzenko made those changes), then you'd have to accomodate for that in setting up the draw call batches in the new stages. It would be doable and wouldn't really be much of a performance problem if you ensure that you only switch the static/dynamic buffers once, but it does make the code more tedious.